server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java

```diff
@@ -135,157 +148,166 @@ protected void masterOperation(
+                        shardsAcked -> finalListener.onResponse(new CreateIndexResponse(true, shardsAcked, indexName)),
                         finalListener::onFailure
                     );
                 } else {
                     finalListener.onResponse(new CreateIndexResponse(false, false, indexName));
                 }
             }, finalListener::onFailure);
-            // TODO: move this to custom class instead of AckedClusterStateUpdateTask
-            AckedClusterStateUpdateTask clusterTask = new AckedClusterStateUpdateTask(Priority.URGENT, request, listener) {
+            CreateIndexTask clusterTask = new CreateIndexTask(request, listener, indexNameRef);
+            clusterService.submitStateUpdateTask("auto create [" + request.index() + "]", clusterTask, clusterTask, executor, clusterTask);
+        }
 
-                @Override
-                public ClusterState execute(ClusterState currentState) throws Exception {
-                    final SystemDataStreamDescriptor dataStreamDescriptor = systemIndices.validateDataStreamAccess(
-                        request.index(),
-                        threadPool.getThreadContext()
-                    );
-                    final boolean isSystemDataStream = dataStreamDescriptor != null;
-                    final boolean isSystemIndex = isSystemDataStream == false && systemIndices.isSystemIndex(request.index());
-                    final ComposableIndexTemplate template = resolveTemplate(request, currentState.metadata());
-                    final boolean isDataStream = isSystemIndex == false
-                        && (isSystemDataStream || (template != null && template.getDataStreamTemplate() != null));
-
-                    if (isDataStream) {
-                        // This expression only evaluates to true when the argument is non-null and false
-                        if (isSystemDataStream == false && Boolean.FALSE.equals(template.getAllowAutoCreate())) {
-                            throw new IndexNotFoundException(
-                                "composable template " + template.indexPatterns() + " forbids index auto creation"
-                            );
-                        }
+        @Override
+        protected ClusterBlockException checkBlock(CreateIndexRequest request, ClusterState state) {
+            return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_WRITE, request.index());
+        }
 
-                        CreateDataStreamClusterStateUpdateRequest createRequest = new CreateDataStreamClusterStateUpdateRequest(
-                            request.index(),
-                            dataStreamDescriptor,
-                            request.masterNodeTimeout(),
-                            request.timeout(),
-                            false
+        private final class CreateIndexTask extends AckedClusterStateUpdateTask {
```

Comment:
Are we getting much benefit from the `extends AckedClusterStateUpdateTask` here? I think if we dropped that and just implemented `AckedClusterStateTaskListener` directly we wouldn't need to construct the `listener` up front and could probably solve the `TODO` about deduplicating the listeners.

Reply:
++ I agree, the main motivation of this PR is to unblock some broken tests quickly (also in ML QA). Are we good doing that kind of refactoring in a follow-up? :)

Reply:
Sure, maybe just leave a TODO here tho.

Reply:
TODO added!